evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,147 @@
+from collections import (
+    defaultdict,
+)
+
 import logging
 
+from cytoolz import (
+    pipe,
+)
+
+import rlp
+
 from web3.contract import (
     Contract,
 )
 
 from eth_utils import (
+    event_signature_to_log_topic,
     is_canonical_address,
     to_checksum_address,
     to_dict,
+    to_tuple,
+)
+
+from evm.rlp.sedes import (
+    address,
+    hash32,
 )
 
 from evm.utils.hexadecimal import (
+    encode_hex,
     decode_hex,
 )
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
 
 from evm.vm.forks.sharding.config import (
     get_sharding_config,
 )
 
 
+class NextLogUnavailable(Exception):
+    pass
+
+
+@to_dict
+def parse_collation_added_data(data_hex):
+    data_bytes = decode_hex(data_hex)
+    score = big_endian_to_int(data_bytes[-32:])
+    is_new_head = bool(big_endian_to_int(data_bytes[-64:-32]))
+    header_bytes = data_bytes[:-64]
+    # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+    sedes = rlp.sedes.List([
+        rlp.sedes.big_endian_int,
+        rlp.sedes.big_endian_int,
+        hash32,
+        hash32,
+        hash32,
+        address,
+        hash32,
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
-    def __init__(self, *args, default_privkey, **kwargs):
+    def __init__(self, *args, log_handler, default_privkey, **kwargs):
+        self.log_handler = log_handler
         self.default_privkey = default_privkey
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         self.config = get_sharding_config()
+        # shard_id -> list
+        # older <---------------> newer
+        self.new_logs = defaultdict(list)
+        # shard_id -> list
+        # newer <---------------> older
```

Comment:
I think this isn't accurate.

- Lets say we get logs `1, 2, 3, 4, 5`
- Log 1 is the oldest
- Log 5 is the newest
- They come in as two batches, `1, 2, 3` and then when we query again after some blocks, `4, 5`

In this case, I think that `unchecked_logs` will be `3, 2, 1, 5, 4`.  Maybe I'm reading the code wrong...?

Reply:
I think you're right. It is not accurate. I will remove this comment.

Reply:
For the algorithm `fetch_candidate_head`, it is from https://github.com/ethereum/sharding/blob/develop/docs/doc.md#fetch-candidate-heads-in-reverse-sorted-order .
It's idea is
> The idea is that this algorithm is guaranteed to check potential head candidates in highest-to-lowest sorted order of score, with the second priority being oldest to most recent.

I am thinking one example, which follows what you assume above.
Assume we have `batch1 = [1, 2, 3]` and `batch2 = [4, 5]`.
First, `batch1` arrives and we have `new_logs = [1, 2, 3]`.
Before `batch2` arrives, the outcome of `fetch_candidate_head` will be `3`, `2`, `1`. Anytime as long as `batch2` arrives, the result of `fetch_candidate_head` would be the  element in `batch2`, instead of the one in `batch1`.
So I think for this case the algorithm is right.

However, if we assume the `batch1 = [1, 1]` and `batch2 = [2]`. As long as batch2 arrives after `batch1` and before `fetch_candidate_head` is called, the order of the outcome would be `1, 1, 2`, which seems not consistent with the idea of the algorithm "The idea is that this algorithm is guaranteed to check potential head candidates in highest-to-lowest sorted order of score, with the second priority being oldest to most recent.".
Example code is in [here](https://gist.github.com/mhchia/dfa9961de7c3fd59f08afaf97552883a)
About this I think I should ask @vbuterin for its detail.

Reply:
The intended principle is that you want to get logs in reverse logical order, so latest block to oldest block, and within a block last transaction to first transaction, and within a transaction last to first created.